File: data-cleaning.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set visual style
sns.set_theme(style="whitegrid")
plt.rcParams["figure.figsize"] = (10, 6)

# -------------------------------------------------------------
# 1. LOAD & INSPECT DATA
# -------------------------------------------------------------
# Using Kaggle IMDb 5000 / metadata standard schema
df = pd.read_csv("movie_metadata.csv")

print("Initial Shape:", df.shape)
print("\nMissing Values per Column:\n", df.isnull().sum()[df.isnull().sum() > 0])

# -------------------------------------------------------------
# 2. DATA CLEANING & PREPARATION
# -------------------------------------------------------------
# Select relevant predictor columns and target (imdb_score)
selected_cols = [
    "movie_title", "title_year", "duration", "budget", "gross",
    "num_voted_users", "num_critic_for_reviews", "director_facebook_likes",
    "cast_total_facebook_likes", "genres", "imdb_score"
]
df_clean = df[selected_cols].copy()

# Remove duplicates based on movie title and release year
df_clean.drop_duplicates(subset=["movie_title", "title_year"], inplace=True)

# Drop rows missing the target variable
df_clean.dropna(subset=["imdb_score"], inplace=True)

# Impute numeric missing values with median
numeric_cols = df_clean.select_dtypes(include=[np.number]).columns
for col in numeric_cols:
    if df_clean[col].isnull().sum() > 0:
        df_clean[col].fillna(df_clean[col].median(), inplace=True)

# Feature Engineering: Extract primary genre and log-transform skewed financials
df_clean["primary_genre"] = df_clean["genres"].apply(lambda x: str(x).split("|")[0] if pd.notnull(x) else "Unknown")
df_clean["log_budget"] = np.log1p(df_clean["budget"])
df_clean["log_gross"] = np.log1p(df_clean["gross"])

print("Cleaned Dataset Shape:", df_clean.shape)

# -------------------------------------------------------------
# 3. EXPLORATORY DATA ANALYSIS & VISUALIZATION
# -------------------------------------------------------------
# Visual 1: Target Variable Distribution (IMDb Rating)
plt.figure(figsize=(8, 5))
sns.histplot(df_clean["imdb_score"], bins=30, kde=True, color="royalblue")
plt.title("Distribution of IMDb Scores")
plt.xlabel("IMDb Score (1-10)")
plt.ylabel("Frequency")
plt.savefig("python_eda_target_distribution.png", dpi=300, bbox_inches="tight")
plt.show()

# Visual 2: Correlation Heatmap of Numeric Predictors
plt.figure(figsize=(9, 7))
corr_matrix = df_clean[numeric_cols].corr()
sns.heatmap(corr_matrix, annot=True, fmt=".2f", cmap="coolwarm", cbar=True, square=True)
plt.title("Correlation Matrix of Numeric Features")
plt.savefig("python_eda_correlation_matrix.png", dpi=300, bbox_inches="tight")
plt.show()

# Visual 3: Top Genres vs. IMDb Score Boxplot
top_genres = df_clean["primary_genre"].value_counts().nlargest(8).index
plt.figure(figsize=(10, 6))
sns.boxplot(
    data=df_clean[df_clean["primary_genre"].isin(top_genres)],
    x="primary_genre",
    y="imdb_score",
    palette="Set2"
)
plt.title("IMDb Score Distribution by Primary Genre (Top 8 Genres)")
plt.xlabel("Genre")
plt.ylabel("IMDb Score")
plt.xticks(rotation=30)
plt.savefig("python_eda_genre_boxplot.png", dpi=300, bbox_inches="tight")
plt.show()

# Export clean data for modeling phase
df_clean.to_csv("imdb_cleaned_week1.csv", index=False)

File: data-cleaning.R
library(dplyr)
library(stringr)
library(ggplot2)
library(corrplot)

# -------------------------------------------------------------
# 1. LOAD DATA
# -------------------------------------------------------------
df <- read.csv("imdb_top_1000.csv", stringsAsFactors = FALSE)

cat("Initial Dimensions:", dim(df), "\n")

# -------------------------------------------------------------
# 2. DATA CLEANING & PREPARATION
# -------------------------------------------------------------
df_clean <- df %>%
  # Deduplicate
  distinct(Series_Title, Released_Year, .keep_all = TRUE) %>%
  # Fix non-numeric Released_Year values
  mutate(Released_Year = suppressWarnings(as.numeric(Released_Year))) %>%
  filter(!is.na(Released_Year)) %>%
  # Clean Runtime and Gross string formatting
  mutate(
    Runtime_min = as.numeric(str_replace(Runtime, " min", "")),
    Gross_clean = as.numeric(str_replace_all(Gross, ",", "")),
    Primary_Genre = str_trim(sapply(strsplit(as.character(Genre), ","), `[`, 1)),
    Certificate = ifelse(is.na(Certificate) | Certificate == "", "Unknown", Certificate)
  ) %>%
  # Median Imputation for numeric features
  mutate(
    Meta_score = ifelse(is.na(Meta_score), median(Meta_score, na.rm = TRUE), Meta_score),
    Gross_clean = ifelse(is.na(Gross_clean), median(Gross_clean, na.rm = TRUE), Gross_clean)
  ) %>%
  # Log-transforms for skewed predictors
  mutate(
    log_No_of_votes = log1p(No_of_votes),
    log_Gross = log1p(Gross_clean)
  ) %>%
  select(
    Released_Year, Runtime_min, Meta_score, No_of_votes,
    Gross_clean, log_No_of_votes, log_Gross, Primary_Genre,
    Certificate, Director, Star1, IMDB_Rating
  )

cat("Cleaned Dimensions:", dim(df_clean), "\n")

# -------------------------------------------------------------
# 3. EXPLORATORY DATA ANALYSIS & VISUALIZATIONS
# -------------------------------------------------------------

# Visual 1: Target Variable Distribution
p1 <- ggplot(df_clean, aes(x = IMDB_Rating)) +
  geom_histogram(aes(y = after_stat(density)), bins = 20, fill = "#1f77b4", color = "white") +
  geom_density(color = "#d62728", linewidth = 1) +
  theme_minimal() +
  labs(
    title = "Distribution of IMDb Top 1000 Ratings",
    x = "IMDb Rating",
    y = "Density"
  )
print(p1)
ggsave("r_eda_target_dist.png", plot = p1, width = 8, height = 5, dpi = 300)

# Visual 2: Correlation Matrix
numeric_cols <- df_clean %>%
  select(IMDB_Rating, Meta_score, Runtime_min, log_No_of_votes, log_Gross, Released_Year)

corr_matrix <- cor(numeric_cols, use = "complete.obs")

png("r_eda_correlation_matrix.png", width = 800, height = 700)
corrplot(
  corr_matrix,
  method = "circle",
  type = "upper",
  addCoef.col = "black",
  tl.col = "black",
  number.cex = 0.8,
  title = "Correlation Matrix of Numeric Predictors",
  mar = c(0, 0, 1, 0)
)
dev.off()

# Visual 3: Primary Genre vs. IMDB Rating
top_genres <- names(sort(table(df_clean$Primary_Genre), decreasing = TRUE)[1:8])

p3 <- df_clean %>%
  filter(Primary_Genre %in% top_genres) %>%
  ggplot(aes(x = reorder(Primary_Genre, IMDB_Rating, median), y = IMDB_Rating, fill = Primary_Genre)) +
  geom_boxplot(show.legend = FALSE, alpha = 0.8) +
  theme_minimal() +
  coord_flip() +
  labs(
    title = "IMDb Rating by Primary Genre (Top 8 Genres)",
    x = "Primary Genre",
    y = "IMDb Rating"
  )
print(p3)
ggsave("r_eda_genre_boxplot.png", plot = p3, width = 8, height = 5, dpi = 300)

# Export clean CSV
write.csv(df_clean, "imdb_top_1000_cleaned_r.csv", row.names = FALSE)